File: ctdl/DynamicProgramming/XauConDoiXungDaiNhat.cpp
#include<bits/stdc++.h>

using namespace std;

bool dp[1001][1001];
void handle(){
	string s;
	cin >> s;
	int n = s.size();
	memset(dp, false, sizeof(dp));
	for(int i=0; i<n; ++i){
		dp[i][i] = true;
	}
	int res = 1;
	for(int k = 1; k<n; ++k){
		for(int i=0; i<n; ++i){
			int j = i+k;
			if(s[i] == s[j] && k==1){
				dp[i][j] = true;
			}else if(s[i] == s[j] && k > 1){
				dp[i][j] = dp[i+1][j-1];
			}else{
				dp[i][j] = 0;
			}
			if(dp[i][j]){
				res = max(res, k+1);
			}
		}
	}
	cout << res;
}
int main(){
	int t;
	cin >> t;
	while(t--){
		handle();
		cout << endl;
	}
	return 0; 
}

File: Python_Practice/DSA/Dynamic_Programming/XauConDoiXungDaiNhat.py
def handle():
    s = input()
    n = len(s)
    res = 1
    dp = [[0 for _ in range(n)] for _ in range(n)]
    for i in range(n):
            dp[i][i] = 1
    for k in range(1, n):
        for i in range(n-k):
            j = i+k
            if s[i] == s[j] and k == 1:
                dp[i][j] = 1
            elif s[i] == s[j] and k > 1:
                dp[i][j] = dp[i+1][j-1]
            else:
                dp[i][j] = 0
            if dp[i][j]:
                res = max(res, k+1)
    print(res)


t = int(input())
for _ in range(t):
    handle()
